src/currentprocess/terminalsource.rs

```diff
@@ -0,0 +1,231 @@
+use std::io::{self, Result, Write};
+use std::sync::{Arc, Mutex, MutexGuard};
+
+pub(crate) use termcolor::Color;
+use termcolor::{ColorChoice, ColorSpec, StandardStream, WriteColor};
+
+use crate::process;
+
+/// Allows you to write to a terminal or other output stream with a lock.
+/// This is a stand-in for std::io::StdoutLock.
+pub trait WriterLock: Write {}
+
+/// This is a stand-in for std::io::Stdout or std::io::Stderr.
+pub trait Writer: Send + Sync {
+    /// Locks the terminal or other output stream.
+    /// This is a stand-in for std::io::Stdout::lock or std::io::Stderr::lock.
+    fn lock(&self) -> Box<dyn WriterLock + '_>;
+}
+
+/// Satisfies `Writer`, and supports optional coloring
+/// to either of the standard output streams, stdout and stderr.
+pub trait Terminal: Writer {
+    fn fg(&mut self, color: Color) -> io::Result<()>;
+    fn bg(&mut self, color: Color) -> io::Result<()>;
+    fn attr(&mut self, attr: Attr) -> io::Result<()>;
+    fn reset(&mut self) -> io::Result<()>;
+    fn carriage_return(&mut self) -> io::Result<()>;
+}
+
+// -------------- stdout -------------------------------
+
+/// Stand-in for std::io::stdout.
+pub trait StdoutTerminalSource {
```

Comment:
I don't understand why this moves across from filesource.

We have outstanding feature requests for an API (e.g. JSON output on stdout, maybe even an RPC style thing), where we'd want to be operating entirely separately from the terminal layer.

Is it possible to do this:
filesource - unchanged vs master
terminalsource - adds in a terminal() method ?

code that previously used stdout doesn't change
code that previously used term2, uses process().terminal()

?

Reply:
I changed it to:
1. `process().stdout()`
2. `process().stdout_terminal()`